RSK `ledger` integration

ahsan-javaiid opened the pull request:
## Description
- This PR adds `ledger` integration for `RSK`
- Currently (main branch) is using hardcoded derivation path [here](https://github.com/tallycash/extension/blob/39b9166cefdc4f41c187fe598ae1f43dd53fff50/background/services/ledger/index.ts#L114) in ledger service. 
- When ledger service connect with RSK app on ledger with eth dpath then `RSK` app will simply reject it.
- So mainly this PR is making derivation path dynamic in ledger service and on UI as well to make it work for `RSK`

## Demo 
<img width="594" alt="Screenshot 2022-11-08 at 6 38 35 PM" src="https://user-images.githubusercontent.com/104683677/200605618-9e5f8827-8483-47d4-b00f-e95b166291e2.png">

## Type of Change
- [X] Fix/ Enhancement / Integration

## Testing information
- Activate `SUPPORT_RSK` env flag
- Connect with ledger
- Activate `RSK` app on ledger
- Good to go
- Transaction signing is working

## Checklist
- [X] Git hooks enabled
- [X] No lint errors
- [X] Git commit is signed

ahsan-javaiid commented:
cc: @0xDaedalus @alepc253

mhluongo commented on ui/pages/Ledger/LedgerImportAccounts.tsx:
```diff
@@ -167,7 +168,10 @@ function LedgerAccountList({
                       <div className="balance">
-                        {balance} {selectedNetwork.baseAsset.symbol}
+                        {balance}{" "}
+                        {appInfo?.appName === "RSK"
```
I don't think this is the approach we want here, otherwise we'd do the same thing for Polygon.

mhluongo replied:
Instead, we should make sure that the selected network and Ledger app match. If they don't, we should ask the user if they'd like to switch to the RSK network... or tell them to switch Ledger apps (to either Ethereum or Polygon, based on the selected network).

This edge case probably needs some design thinking @VladUXUI

VladUXUI replied:
Can i get a little bit more context here?
Is this part of the onboarding or something else? asking because in onboarding we don't currently have option for users to change networks (although it's on the way).

jagodarybacka replied:
Just for context - Polygon app on Ledger is really bad (and seems to be just a fork/extension of Ethereum app), was throwing weird errors and it was just a more stable approach to encourage users to use the Ethereum app for both ETH and Polygon. Maybe Rsk app is more stable.

alepc253 replied:
@jagodarybacka the problem is that you have to use the right app with the right network because the Ledger app only works with the right derivation path (ETH/Polygon and RSK have different dpath)

ahsan-javaiid replied:
The problem is that we do not have any UI to select network/derivation path when user clicks on "connect to ledger" button. Tally wallet is assuming `ethereum` network by default [here](https://github.com/tallycash/extension/blob/39b9166cefdc4f41c187fe598ae1f43dd53fff50/background/services/ledger/index.ts#L114) and attempt the connect with ledger using `ethereum` derivation path. In case of other networks like `RSK` connection get rejected by ledger app (RSK in this case) because tally wallet is passing hardcoded derivation path [here](https://github.com/tallycash/extension/blob/39b9166cefdc4f41c187fe598ae1f43dd53fff50/background/services/ledger/index.ts#L114)

I can think of below two approaches to solve this issue:
**Approach 1**: Auto detect which app is running on device and use dPath w.r.t that app
Like in this PR i am auto detecting the active app on ledger and then passing dPath dynamically according to connected app. 

**Approach 2**: Provide a UI to select network or derivation path before connecting to ledger so that user can select right network according to running app on ledger device. 

**So when user clicks on "connect to ledger" button, then user will be taken to derivation path selection screen instead of directly connecting to ledger as its doing right now and assuming `ethereum` derivation path**

IMO, second approach is better and i can update the pull request with this approach. 

Most welcome in case of any other thought's 👍

jagodarybacka replied:
> Provide a UI to select network or derivation path before connecting to ledger

@VladUXUI I remember you were thinking about something similar before right?

0xDaedalus replied:
Follow-up question both for @alepc253 and @mr-michael - Are we OK turning on RSK without this PR being merged?

alepc253 replied:
@0xDaedalus yes, we are OK. This feature is important but can be supported in next releases. Launching RSK asap is the main goal.

ahsan-javaiid replied:
Ready for review.

mhluongo reviewed (dismissed):
From [`@zondax/ledger-icp`](https://github.com/Zondax/ledger-icp)'s docs...

> Please:
> * Do not use in production
> * Do not use a Ledger device with funds for development purposes.
> * Have a separate and marked device that is used ONLY for development and testing

We should aim to do this without additional outside dependencies, especially not experimental dependencies.

ahsan-javaiid commented:
Tally Team 👋

I have updated the pull request with `ui` based approach. Now before connecting to `ledger` user will be able to select dPath w.r.t ledger app. 

See Demo video here: 

https://user-images.githubusercontent.com/104683677/201657410-29363c75-9101-461d-97a6-33bad3af7c9e.mp4



- Also resolved previous feedback and not using any dependencies
- Ready for review and most welcome for any thoughts 🤔

greg-nagy commented on background/constants/networks.ts:
```diff
@@ -85,7 +85,7 @@ export const TEST_NETWORK_BY_CHAIN_ID = new Set(
 )
 
-export const NETWORK_FOR_LEDGER_SIGNING = [ETHEREUM, POLYGON]
+export const NETWORK_FOR_LEDGER_SIGNING = [ETHEREUM, POLYGON, RSK]
```
We have renamed `RSK` > `ROOTSTOCK` in https://github.com/tallycash/extension/pull/2625
and this breaks the build and also the tests.

Sorry for this issue! 

Are you ok fixing this, or would you prefer me to do it? 

The fix patch is attached.
[RSKtoROOTSTOCK.patch.txt](https://github.com/tallycash/extension/files/10014510/RSKtoROOTSTOCK.patch.txt)

ahsan-javaiid replied:
Done

greg-nagy commented on background/redux-slices/migrations/to-6.ts:
```diff
@@ -23,6 +23,7 @@ export default (prevState: Record<string, unknown>): NewState => {
       currentDeviceID: null,
       devices: {},
       usbDeviceCount: 0,
+      derivationPath: null,
```
Probably not necessary but ❤️ the details

ahsan-javaiid replied:
Done

VladUXUI commented:
This is looking awesome! 👏 😍 
We will probably update soon with new onboarding, but i'm happy for this to be shipped as is (ui/ux wise)

ahsan-javaiid commented:
Hi Tally team 👋

Feel free to share if there are any further thoughts on this PR.

Shadowfiend commented on ui/pages/Ledger/LedgerSelectDerivationPath.tsx:
```diff
@@ -0,0 +1,55 @@
+import { useBackgroundDispatch } from "../../hooks"
+import SelectDerivationPath from "../../components/Onboarding/OnboardingDerivationPathSelect"
+
+const DERIVATION_PATH_TO_NETWORK: { [key: string]: EVMNetwork } = {
+  "m/44'/137'/0'/0": ROOTSTOCK,
+  "m/44'/60'/0'/0": ETHEREUM,
+  "m/44'/1'/0'/0": ETHEREUM,
+  "m/44'/60'/0'": ETHEREUM,
+}
```
Feels like the ledger lib should have some way to resolve this somewhere...

ahsan-javaiid replied:
Thanks for the suggestion @Shadowfiend. Purpose of this mapping is get the right network against the derivation path that is returned by derivation path dropdown. [Here](https://github.com/ahsan-javaiid/extension/blob/d69c09484a9c423af50256aab562160c44d0421c/ui/pages/Ledger/LedgerSelectDerivationPath.tsx#L28) i am setting the network selected by the user. 
Objective here is to set the right network based on derivation path selected by user.  

```
Feels like the ledger lib should have some way to resolve this somewhere...
```

I looked in tally code base (ledger related files) to find something as you suggested. But i think such mapping or something related is not there unless there is some other intelligent way of achieving this. 

Would be great if you can share some hint 🙏

ahsan-javaiid replied:
```
Feels like the ledger lib should have some way to resolve this somewhere...
```
@Shadowfiend Resolved this concern

ahsan-javaiid commented:
Hi Tally team 👋

Updated this pull request and implemented the `ledger upgrade flow to include dropdown`

Reference issue: https://github.com/tallycash/extension/issues/2623

Demo:

<img width="618" alt="Screenshot 2022-12-06 at 11 13 46 AM" src="https://user-images.githubusercontent.com/104683677/205837411-c1617f60-79c4-4714-bff6-11036eeb3974.png">


Demo video:

https://user-images.githubusercontent.com/104683677/205837481-d63cbfe9-fd77-4a8c-8405-2ecb4a6f1066.mov


Feel free to review the pull request and share thought's if anything needs to changed. 🙏

kkosiorowska commented on ui/components/Onboarding/OnboardingDerivationPathSelect.tsx:
```diff
@@ -51,11 +52,16 @@ export default function OnboardingDerivationPathSelect({
+  const selectedIndex = initialDerivationPaths.findIndex(
+    (path) => path.value === `m/${ledgerState.derivationPath}`
+  )
+  const [defaultIndex, setDefaultIndex] = useState<number>(
+    selectedIndex === -1 ? 0 : selectedIndex
+  )
```
```suggestion
  const [defaultIndex, setDefaultIndex] = useState(
    selectedIndex === -1 ? 0 : selectedIndex
  )
```

ahsan-javaiid replied:
Done 👍

kkosiorowska commented on ui/components/LedgerMenu/LedgerMenuListItemChild.tsx:
```diff
@@ -0,0 +1,28 @@
+import { EVMNetwork } from "@tallyho/tally-background/networks"
+
+interface Props {
+  network: EVMNetwork
```
I think that we don't need to the whole object here. The network name is enough for us

ahsan-javaiid replied:
Done 👍

kkosiorowska commented on ui/components/LedgerMenu/LedgerMenuListItemChild.tsx:
```diff
@@ -0,0 +1,28 @@
+  network: EVMNetwork
+}
+
+export default function LedgerMenuListItemChild(props: Props): ReactElement {
```
`LedgerMenuListItemChild` name is a little misunderstood for me.  What do you think about  `EcosystemNetworkIcon` ?  Maybe someone else has some more ideas.

ahsan-javaiid replied:
Done 👍

kkosiorowska commented on ui/components/LedgerMenu/LedgerMenuProtocolListItem.tsx:
```diff
@@ -0,0 +1,148 @@
+import { ETHEREUM, ROOTSTOCK } from "@tallyho/tally-background/constants"
+import LedgerMenuListItemChild from "./LedgerMenuListItemChild"
+
+const DEFAULT_DERIVATION_PATH = "44'/60'/0'/0/0"
+const ROOTSTOCK_DERIVATION_PATH = "44'/137'/0'/0"
```
We declare these values in several places. Let's make it const so that if there is a change, it takes place in one place.

ahsan-javaiid replied:
Just a suggestion to make it part of network. Let me know if this idea makes sense or should i move it somewhere else?

kkosiorowska replied:
hmm I think it can be a part of the network but probably we don't want to mix this thing.  Let's wait for @0xDaedalus  opinion.

kkosiorowska commented on ui/components/LedgerMenu/LedgerMenuProtocolListItem.tsx:
```diff
@@ -0,0 +1,148 @@
+    <li
+      className={classNames({ select: isSelected })}
+      onClick={() => {
+        dispatch(setSelectedNetwork(network))
+        if (network.chainID === ROOTSTOCK.chainID) {
+          dispatch(derivationPathChange(ROOTSTOCK_DERIVATION_PATH))
+        } else {
+          dispatch(derivationPathChange(DEFAULT_DERIVATION_PATH))
+        }
```
Let's wrap it up in a handle function

ahsan-javaiid replied:
Done 👍

kkosiorowska reviewed (changes requested):
I left some comments on the code. 

@VladUXUI Could you check the UI part? The solution is a little different from the [design](https://github.com/tallycash/extension/issues/2623)

ahsan-javaiid commented on background/constants/networks.ts:
```diff
@@ -114,7 +114,7 @@ export const TEST_NETWORK_BY_CHAIN_ID = new Set(
 )
 
-export const NETWORK_FOR_LEDGER_SIGNING = [ETHEREUM, POLYGON]
+export const NETWORK_FOR_LEDGER_SIGNING = [ETHEREUM, POLYGON, ROOTSTOCK]
```
Done 👍

ahsan-javaiid commented on background/main.ts:
```diff
@@ -1011,6 +1011,10 @@ export default class Main extends BaseService<never> {
       this.store.dispatch(setUsbDeviceCount({ usbDeviceCount }))
     })
+
+    uiSliceEmitter.on("derivationPathChange", async (path: string) => {
```
Done 👍

ahsan-javaiid commented on ui/pages/Ledger/LedgerSelectNetwork.tsx:
```diff
@@ -0,0 +1,40 @@
+  const { t } = useTranslation("translation", {
+    keyPrefix: "ledger.onboarding.selectLedgerApp",
+  })
+  const buttonLabel = t("continueButton")
```
Done 👍

ahsan-javaiid commented on ui/components/LedgerMenu/LedgerMenuProtocolListItem.tsx:
```diff
@@ -0,0 +1,148 @@
+        </div>
+      </div>
+      <div className="right">
+        <div className="title">{formatNetworkName()}</div>
```
Done 👍

ahsan-javaiid commented on ui/components/Onboarding/OnboardingDerivationPathSelect.tsx:
```diff
@@ -51,11 +52,16 @@ export default function OnboardingDerivationPathSelect({
   const { t } = useTranslation("translation", { keyPrefix: "onboarding" })
   const [derivationPaths, setDerivationPaths] = useState(initialDerivationPaths)
-
+  const ledgerState = useBackgroundSelector((state) => state.ledger)
```
Done 👍

VladUXUI commented:
This is looking good! 
Will build today and test it out, as i see some small UI clean-up needed.
We since added BSC and Avalanche, so this will probably need to be updated a bit.
The logic is quite simple i hope, if it has a Ledger app, then it need a different select

VladUXUI commented:
Hey hey, so i just tested out the network selector from a ui pov, and added a few clean-up things. Pasting this as an image as it was easier for me to do.
![image](https://user-images.githubusercontent.com/80846886/207243196-e7e73936-1501-4696-9f44-349f2d56a977.png)


Also i moved the Figma to a public file so you can access it and look at stuff.
https://www.figma.com/file/EiXpBpNz2HcoeFJftgC5oc/Tally-Public?node-id=1209%3A2200

Do let me know if it isn't working

VladUXUI commented:
The next page after network is selected needs a back button, just in case users change their mind

ahsan-javaiid commented:
Update: Design improvements added 👍

VladUXUI commented:
Looking good @ahsan-javaiid ! 
The two remaining items are:
- adding Binance & Avalanche as selection
- Adding a back button on Ledger connect page

Are these things that you would like to do? or should i ask the team to create another PR for these?

ahsan-javaiid commented:
Update:
- Added new chains selection 

<img width="540" alt="Screenshot 2022-12-13 at 7 34 08 PM" src="https://user-images.githubusercontent.com/104683677/207361366-72275065-a605-4b9c-85db-fb20837dfcdd.png">

- Added back button 

<img width="684" alt="Screenshot 2022-12-13 at 7 35 29 PM" src="https://user-images.githubusercontent.com/104683677/207361730-49726816-e2c6-406f-8035-e3aa379f6b11.png">


Let me know if anything needs to be changed further!

VladUXUI commented:
Looking very good on my end! Awesome job @ahsan-javaiid!
Will let somebody from dev team to give final say

kkosiorowska commented:
It looks great after the design improvements! :grinning:  But I have a small problem with the AVAX network. I cannot connect by selecting this network. I don't have this issue for RSK or BNB chain. 

![Screenshot 2022-12-15 at 09 20 13](https://user-images.githubusercontent.com/23117945/207809994-9ed656a2-1569-46ce-98b7-39324626acfa.png)

kkosiorowska reviewed (commented):
We're really close 🚀  I left a few tiny comments.  

Let's still wait for @0xDaedalus  final feedback. 😀 
I wonder about one more thing. BSC and AVAX have feature flags.  In this case, should we add the ability to switch off/on these networks?

ahsan-javaiid commented on background/redux-slices/ledger.ts:
```diff
@@ -58,6 +59,7 @@ export const initialState: LedgerState = {
   currentDeviceID: null,
   devices: {},
   usbDeviceCount: 0,
+  derivationPath: undefined,
```
Done 👍

ahsan-javaiid commented on ui/components/LedgerMenu/LedgerMenuProtocolListItem.tsx:
```diff
@@ -0,0 +1,161 @@
+            ? t("ecosystem", { network: network.name })
+            : network.name}
+        </div>
+        <div className="sub_title">{ledgerAppSubTitle(ecosystem)}</div>
```
Done 👍

ahsan-javaiid commented:
Hi Tally team 👋

Users are eager to use Tally wallet with Roostock <> Ledger. I would be happy to do any further changes to get this pull request considered.

Feel free to share further thoughts on this pull request.

alepc253 commented:
hey @kkosiorowska  @jagodarybacka, any update on this feature? is there anything else we could help with from Rootstock team?

jagodarybacka commented on background/services/ledger/index.ts:
```diff
@@ -113,15 +114,24 @@ type Events = ServiceLifecycleEvents & {
 
 export const idDerivationPath = "44'/60'/0'/0/0"
 
+const ROOTSTOCK_DERIVATION_PATH = "44'/137'/0'/0"
```
We can use [this](https://github.com/tallyhowallet/extension/blob/main/background/constants/networks.ts#L27) instead

ahsan-javaiid replied:
Done!

jagodarybacka commented on background/services/ledger/index.ts:
```diff
@@ -113,15 +114,24 @@ type Events = ServiceLifecycleEvents & {
 
 export const idDerivationPath = "44'/60'/0'/0/0"
```
This should be moved to consts and replaced in multiple places

ahsan-javaiid replied:
Done!

jagodarybacka commented on ui/components/LedgerMenu/LedgerMenuProtocolListItem.tsx:
```diff
@@ -0,0 +1,167 @@
+      role="presentation"
+    >
+      <div className="left">
+        <div className="icon_wrap">
+          <span className="icon" />
```
to be replaced with SharedNetworkIcon

ahsan-javaiid replied:
Done!

jagodarybacka reviewed (changes requested):
Sorry for the delay 😭 Awesome job was done here!

From UI testing:
- I think this selector should be hidden for now as it is bugged and is not really changing dpath but is going back to network selector sceen
![image](https://user-images.githubusercontent.com/20949277/217542688-76cdc4b9-ded8-464b-9a43-6bb18cc9414b.png)
- with tabbed onboarding turned on `SUPPORT_TABBED_ONBOARDING=true` we should support a similar flow cc @VladUXUI 🤔 but it shouldn't block this PR IMO

A couple of comments, mostly non-blocking, besides hiding bugged dpath selector

ahsan-javaiid commented on ui/_locales/en/messages.json:
```diff
@@ -113,7 +120,8 @@
         "step2": "Enter pin to unlock",
-        "step3": "Open Ethereum App"
+        "step3": "Open {{network}} App",
+        "derivationPath": "Select derivation path to connect with ledger"
```
Done!

ahsan-javaiid commented on ui/components/LedgerMenu/EcosystemNetworkIcon.tsx:
```diff
@@ -0,0 +1,27 @@
+  networkName: string
+}
+
+export default function EcosystemNetworkIcon(props: Props): ReactElement {
```
Done!

ahsan-javaiid commented:
> Sorry for the delay 😭 Awesome job was done here!
> 
> From UI testing:
> 
> * I think this selector should be hidden for now as it is bugged and is not really changing dpath but is going back to network selector sceen
>   ![image](https://user-images.githubusercontent.com/20949277/217542688-76cdc4b9-ded8-464b-9a43-6bb18cc9414b.png)
> * with tabbed onboarding turned on `SUPPORT_TABBED_ONBOARDING=true` we should support a similar flow 🤔 but it shouldn't block this PR IMO
> 
> A couple of comments, mostly non-blocking, besides hiding bugged dpath selector

Done!

VladUXUI commented:
With the addition of network selection at the beginning of the screen. Do we still need dpath selection?
Curious how many users have custom dpath in Ledger

jagodarybacka commented:
@VladUXUI I would say we do need it for the convenience of more advanced users, it would be a nice improvement here. For seed path import there is no way to choose the network there so we can either do it based on dpath selector or network selector 🤔 

@ahsan-javaiid ty for fixes! 

@0xDaedalus could you take a look - I think it's good to go and we can take care of tabbed onboarding later

VladUXUI commented:
Yeah, agreed. But i would rather fix than hide?
It looks like it's fixed

ahsan-javaiid commented:
Hi team: 👋

Just a reminder to consider this pull request.🎗 Feel free to share if anything blocking on this pull request. 🚧

Hope to get this PR considered soon. 🙏

jagodarybacka commented:
Hey @ahsan-javaiid I checked it yesterday and overall it looks nice. One thing I noticed we missed is that now we are missing Ledger live dpath when Ethereum is selected while importing Ledger. We will figure that out with @VladUXUI.

We are planning to release tabbed onboarding soon so we think its best to merge this PR after the initial release of tabbed onboarding is done 🔜 
So no action is needed on your side right now, we will take care of it 🫶

Shadowfiend commented:
Hey folks, I wanted to poke at what is now #3385 before going deep on review here. Now that 3385 is written up, I'm going to try and give this some deep 👀 so we can get over the finish line 💪
